src/stats/man/rankrel.1
.TH RANKREL 1 "January 20, 1987" "\(co 1987 Gary Perlman" "|STAT" "UNIX User's Manual"
.SH NAME
rankrel \- rank order statistics for related samples
.SH SYNOPSIS
.B rankrel
[-ry] [-c maxcases] [names]
.SH DESCRIPTION
.I rankrel
analyses data from ordinally ranked data obtained from related/matched samples.
The input consists of scores from several samples, conditions, or groups.
Each condition's data are in a separate column.
The scores need not be ranks; they will be ranked by the program.
For each condition, the number of scores, extrema and quartiles are reported.
Conditions are compared for equality of location using
the sign test,
the Wilcoxon signed-ranks test,
and the Friedman two-way analysis of variance of ranks.
A matrix of Spearman rank-order correlation coefficients (rho) is printed.
.PP
The sign test and the Wilcoxon test are only used when there are two
conditions.
When there are fewer than 25 paired cases that are different,
the exact binomial probability is computed;
for larger N, the normal approximation is used.
.SS "Probability of Obtained Statistics
Functions computing exact probabilities of Wilcoxon and Friedman
could not be found when the program was written,
so for small samples,
statistical tables at the back of a text should be consulted.
For large samples,
normal and Chi-square approximations are adequate.
.SH OPTIONS
.de OP
.TP
.B -\\$1 \\$2
..
.OP c maxcases
Set the maximum number of input cases.
Use the -L option to see the default.
.OP r
Request a report of average ranks for conditions.
.OP s
Stop significance tests from printing.
This option is useful when the Spearman rho values are of primary interest.
.OP y
When computing the normal approximation
for the probability of the Wilcoxon test,
Yates' correction for continuity is applied.
This option stops its use.
There are no cases where Yates' correction should not be used,
but the option is useful to check textbook examples for accuracy.
.br
.if t .ne 2i
.SH EXAMPLE
.PP
The following data are from Siegel, page 79.
The command names the conditions "school" and "home."
.nf
	> rankrel school home
	82	63
	69	42
	73	74
	43	37
	58	51
	56	43
	76	80
	65	62
.if n .sp
.if t .sp 6p
.ta .5i +.5i
sign test: 0.144531 (one-tailed)
Wilcoxon T = 4, N = 8
	p approximated with z: .026892 (one-tailed)
	tabled critical value for T for one-tailed p = .025: 4
Friedman R = 2
Spearman Rank Correlation (rho) = .786
.fi
.SH LIMITS
Use the -L option to determine the program limits.
.SH "MISSING VALUES
Cases with missing data values (NA) are counted but not included in the analysis.
.SH "SEE ALSO
pair(1), regress(1), and anova(1)
perform the normal-theory parametric counterparts
to this non-parametric, distribution-free analysis.
To see a scatterplot of ranks,
the ranksort(1) filter can be used as a pre-processor for
the pair(1) plotting option.
rankind(1) analyses ordinal data for independent conditions.
.sp
Siegel, S. (1956)
.ul
Nonparametric Statistics for the Behavioral Sciences.
New York: McGraw-Hill.
.SH WARNING
When the program advises to check a table for exact probabilities of
significance tests,
it may still compute approximate values.
These approximations should not be used for serious work.
.ig
             N      Min      25%   Median      75%      Max
school       8    43.00    57.00    67.00    74.50    82.00
home         8    37.00    42.50    56.50    68.50    80.00
Total       16    37.00    47.00    62.50    73.50    82.00

Binomial Sign Test:
	Number of cases school is above home:   6
	Number of cases school is below home:   2
	One-tailed probability (exact)          0.144531

Wilcoxon Matched-Pairs Signed-Ranks Test:
    Comparison of school and home
	T                                       4.000000
	N (number of differences)               8.000000
	z                                       1.928571
	One-tailed probability                  0.026892
	NOTE: Yates' correction for continuity applied
	Check a table for T with N = 8

Friedman Chi-Square Test for Ranks:
	Chi-square of ranks                     2.000000
	chisq       2.000000     df   1      p  0.157299
	Check a table for Friedman with N = 8

Spearman Rank Correlation (rho) [corrected for ties]:
	Check a table for Spearman rho with N = 8
  school     home 
            0.786 school
   0.786          home
..


src/stats/src/rankrel.c
/* Copyright 1987 Gary Perlman */

#include "stat.h"

PGM(rankrel,Rank Order Analysis for Related Conditions,5.2,02/04/87)

#ifdef __STDC__
double cor (float *x, float *y, int n);
double critcor (int n, double p);
int ranksort (float *vec, float *svec, int *ivec, int n);
double	pof (double F, int df1, int df2);
#else
double cor ();         /* compute correlation */
double critcor ();     /* compute critical correlation value */
int ranksort ();       /* ranksort items */
double	pof ();        /* prob of F ratio */
#endif

/*
	matched groups median test
*/

/*
	rankrel centile getopt setreal number prodlist \
		primes chisq z setint ranksort ordstat
*/
#define	PFVEC(vec,count) \
	{ \
	int 	ind; \
	for(ind=0;ind<count;ind++) \
		printf("%g ", vec[ind]); \
	printf ("\n"); \
	}

/* critical N's over which to use approximations */
#define	NWILCOXON    16      /* according to Snedecor & Cochran */
#define	NSIGNTEST    25      /* according to Siegel */
#define	NSPEARMAN    10      /* according to Siegel */

#define	MAXCOND	     20       /* maximum number of conditions */
#define MAXCHARS BUFSIZ       /* maximum number of chars in lines */
#define	MAXDATA     100       /* max number of input data cases */
int 	Maxdata = MAXDATA;

char	*Condname[MAXCOND];   /* condition names */
float	*Condat[MAXCOND];     /* pointers into all data for each condition */
int 	Nconds;               /* number of conditions == ncols */
int 	Count;                /* number of data in each condition */
int 	NAcount[MAXCOND];     /* number of NA missing values */
int 	TNAcount;             /* number of missing cases */

/* OPTIONS */
Boole	InfoVersion;          /* print version information */
Boole	InfoLimits;           /* print program limits */
Boole	InfoOptions;          /* print usage information */
Boole	Rankmeans;            /* print average ranks for conditions? */
Boole	Yates = TRUE;         /* apply Yates' correction for continuity */
Boole	Docomps = TRUE;       /* do comparison tests? **/
#define	noteyates()	printf("\tNOTE: Yates' correction for continuity applied\n")

#define	statheader(header) printf ("\n%s:\n", header)

/*FUNCTION main: rankrel */
main (argc, argv) char **argv;
	{
	ARGV0;
	initial (argc, argv);
	checkstdin ();
	readdata ();
	summarize ();
	if (Nconds > 1)
		{
		if (Docomps)
			{
			if (Nconds == 2)
				{
				dosigntest (0, 1);
				dowilcoxon (0, 1);
				}
			dofriedman ();
			}
		dospearman ();
		}
	exit (SUCCESS);
	}

/*FUNCTION initial: set options and condition names */
initial (argc, argv) char **argv;
	{
	extern	char	*optarg;
	extern	int 	optind;
	int 	errflg = 0;
	int 	C;
	int 	cond;
	while ((C = getopt (argc, argv, "c:rsyLOV")) != EOF)
		switch (C)
			{
			case 'O': InfoOptions = TRUE; break;
			case 'V': InfoVersion = TRUE; break;
			case 'L': InfoLimits = TRUE; break;
			case 'c':
				if (setint (Argv0, 'c', optarg, &Maxdata, 1, MAXINT))
					errflg++;
				break;
			case 'r':
				Rankmeans = TRUE;
				break;
			case 's':
				Docomps = FALSE;
				break;
			case 'y':
				Yates = FALSE;
				break;
			default: errflg++; break;
			}
	if (errflg)
		USAGE ("[-ry] [-c maxcases] [names]")
	usinfo ();
	for (cond = 0; optind + cond < argc; cond++)
		{
		if (cond >= MAXCOND)
			ERRMANY (condition names,MAXCOND)
		Condname[cond] = argv[optind+cond];
		}
	}

/*FUNCTION usinfo: online help */
usinfo ()
	{
	if (InfoVersion)
		pver (Version);
	if (InfoLimits)
		{
		plim (Argv0);
		const (MAXCOND,  "maximum number of conditions");
		const (Maxdata,  "maximum number of input data points");
		const (MAXCHARS, "maximum number of characters in input lines");
		}
	if (InfoOptions)
		{
		ppgm (Argv0, Purpose);
		iopt ('c', "maxcases", "maximum number of input cases", Maxdata);
		lopt ('r', "print average ranks for conditions", Rankmeans);
		lopt ('s', "run comparative significance tests", Docomps);
		lopt ('y', "use Yates' correction for continuity", Yates);
		oper ("[names]", "condition names", "Cond-1 Cond-2 ...");
		}
	if (InfoVersion || InfoLimits || InfoOptions)
		exit (SUCCESS);
	}

/*FUNCTION newname: make a new default name for a condition */
char *
newname (cond)
int 	cond;
	{
	char	*name = myalloc (char, 10);
	(void) sprintf (name, "Cond-%d", cond);
	return (name);
	}

/*FUNCTION checktable: tell analyst to look for value in a table */
checktable (stat, parname, parval)
char	*stat;
char	*parname;
int 	parval;
	{
	printf ("\tCheck a table for %s", stat);
	if (parname && *parname)
		printf (" with %s = %d", parname, parval);
	putchar ('\n');
	}

/*FUNCTION: readdata: read in data into global arrays */
readdata ()
	{
	char	line[MAXCHARS];
	char	*sarray[MAXCOND];
	int 	ncols;
	int 	cond;
	Boole	dropcase;
	
	while (gets (line))
		{
		ncols = parselin (line, sarray, MAXCOND);
		if (ncols == 0)
			continue;
		dropcase = FALSE;
		for (cond = 0; cond < ncols; cond++)
			if (isna (sarray[cond]))
				{
				NAcount[cond]++;
				dropcase = TRUE;
				}
		if (dropcase == TRUE)
			{
			TNAcount++;
			continue;
			}
		if (Nconds == 0) /* initialize */
			{
			Nconds = ncols;
			if (Nconds > MAXCOND)
				ERRMANY (conditions,MAXCOND)
			for (cond = 0; cond < Nconds; cond++)
				{
				Condat[cond] = myalloc (float, Maxdata);
				if (Condat[cond] == NULL)
					ERRSPACE (data)
				}
			}
		if (Count == Maxdata)
			ERRMANY (cases,Maxdata)
		if (ncols != Nconds)
			ERRRAGGED
		for (cond = 0; cond < Nconds; cond++)
			if (number (sarray[cond]))
				Condat[cond][Count] = atof (sarray[cond]);
			else
				ERRNUM (sarray[cond],input value)
		Count++;
		}
	if (Count <= 1)
		ERRDATA
	for (cond = 0; cond < Nconds; cond++)
		if (Condname[cond] == NULL || *Condname[cond] == '\0')
			Condname[cond] = newname (cond+1);
	}

/*FUNCTION summarize: print summary statistics */
summarize ()
	{
	int 	cond;
	int 	i;
	int 	allcount = 0;
	float	*alldata = myalloc (float, Nconds * Count);
	int 	tna = 0;
	
	if (TNAcount)
		printf ("Cases with missing data: %d ignored\n", TNAcount);
	for (cond = 0; cond < Nconds; cond++)
		tna += NAcount[cond];
	for (cond = 0; cond < Nconds; cond++)
		{
		if (alldata)
			for (i = 0; i < Count; i++)
				alldata[allcount++] = Condat[cond][i];
		ordstat (Condat[cond], Count, cond, Condname[cond], NAcount[cond]);
		}
	if (alldata != NULL)
		{
		if (Nconds > 1)
			ordstat (alldata, Nconds*Count, Nconds, "Total", tna);
		free ((char *) alldata);
		}
	else
		WARNING (not enough space to compute grand median)
	}

/*FUNCTION dosigntest: compute sign test for two conditions */
/*
	ties are dropped and N is reduced
	if N > NSIGNTEST, then use normal approximation: apobin
	else use exact binomial probability
*/
dosigntest (cond1, cond2)
int 	cond1, cond2;
	{
	float	*data1 = Condat[cond1];
	float	*data2 = Condat[cond2];
	int 	N = 0;                   /* number of cases in test */
	int 	r = 0;                   /* number of cases first > second */
	int 	i;
	double	p;
	double	apobin (), pobin ();
	
	statheader ("Binomial Sign Test");
	
	for (i = 0; i < Count; i++)
		{
		if (data1[i] == data2[i])
			continue;
		if (data1[i] > data2[i])
			r++;
		N++;
		}
	printf ("\tNumber of cases %s is above %s: %3d\n",
		Condname[cond1], Condname[cond2], r);
	printf ("\tNumber of cases %s is below %s: %3d\n",
		Condname[cond1], Condname[cond2], N-r);
	if (r < N * 0.5)
		r = N - r;
	if (N > NSIGNTEST)
		{
		p = apobin (N, 1, 2, r);
		tprint ("One-tail probability approximation", p);
		noteyates ();
		}
	else
		{
		p = pobin (N, 1, 2, r);
		tprint ("One-tail probability (exact)", p);
		}
	}

/*FUNCTION dowilcoxon: compute Wilcoxon test for two conditions */
/*
	T = min (sumposranks, sumnegranks)
	if N > NWILCOXON, use normal approximation (seems good even when N = 8)
		mean = N (N+1) / 4
		se = sqrt (N (N+1) (2N+1) / 24 )
		z = (T - U) / se      (one tailed)
*/
dowilcoxon (cond1, cond2)
int 	cond1, cond2;
	{
	float	*data1 = Condat[cond1];
	float	*data2 = Condat[cond2];
	double	diff;                      /* data1 - data2 */
	float	*rank;                     /* absolute differences of conditions */
	Boole	*neg;                      /* is difference negative? */
	int 	i;                         /* loop varioable for cases */
	double	sumnegranks = 0.0;         /* sum of all negatively signed ranks */
	double	sumposranks = 0.0;         /* sum of all positively signed ranks */
	double	T;                         /* Wilcoxin statistic */
	int 	N = 0;                     /* number of non-zero differences */
	double	mean;
	double	se;
	double	z, poz ();
	
	rank = myalloc (float, Count);
	neg = myalloc (Boole, Count);
	if (rank == NULL || neg == NULL)
		ERRSPACE(Wilcoxon Test)
	
	statheader ("Wilcoxon Matched-Pairs Signed-Ranks Test");
	printf ("    Comparison of %s and %s\n", Condname[cond1], Condname[cond2]);
	for (i = 0; i < Count; i++)
		{
		diff = data1[i] - data2[i];
		if (!fzero (diff))
			{
			neg[N] = (diff < 0.0);
			rank[N] = fabs (diff);
			N++;
			}
		}
	if (N == 0)
		{
		printf ("\tHmmm, all these pairs seem to be the same.\n");
		return;
		}
	if (ranksort (rank, NULL, NULL, N))
		ERRMSG0 (could not rank data for Wilcoxon test)
	for (i = 0; i < N; i++)
		if (neg[i])
			sumnegranks += rank[i];
		else
			sumposranks += rank[i];
	T = min (sumnegranks, sumposranks);
	tprint ("T (smaller ranksum of like signs)", T);
	tprint ("N (number of signed differences)", (double) N);
	mean = N * (N+1) * 0.25;
	se = sqrt (N * (N+1) * (2.0 * N + 1.0) / 24.0);
	z = fabs (T - mean);
	if (Yates)
		{
		z -= 0.5;
		if (z < 0.0)
			z = 0.0;
		}
	z /= se;
	tprint ("z", z);
	tprint ("One-tail probability approximation", 1.0 - poz (z));
	if (Yates)
		noteyates ();
	if (N <= NWILCOXON)
		checktable ("T", "N", N);
	free ((char *) neg);
	free ((char *) rank);
	}

/*FUNCTION dofriedman: multi-condition test */
dofriedman ()
	{
	double	R = 0.0;            /* sum of squared sum of ranks in conds */
	double	chi_r;              /* function of R and formula */
	double	pochisq ();
	int 	i, cond;
	float	rank[MAXCOND];      /* ranks of each case computed here */
	double	sumrank[MAXCOND];   /* sum of ranks in a condition */
	/* the following two vectors save allocation time for ranksort */
	float	sorted[MAXCOND];    /* sorted condition values */
	int 	order[MAXCOND];     /* sorted order index numbers */
	Boole	needtable = FALSE;  /* should we recommend looking at a table? */
	
	for (cond = 0; cond < Nconds; cond++)
		sumrank[cond] = 0.0;
	for (i = 0; i < Count; i++)
		{
		for (cond = 0; cond < Nconds; cond++)
			rank[cond] = Condat[cond][i];
		if (ranksort (rank, sorted, order, Nconds))
			ERRMSG0 (could not rank data for Friedman test)
		for (cond = 0; cond < Nconds; cond++)
			sumrank[cond] += rank[cond];
		}
	if (Rankmeans)
		{
		statheader ("Average Ranks");
		for (cond = 0; cond < Nconds; cond++)
			printf ("\t\t%-12s %10.1f %10.2f\n",
				Condname[cond], sumrank[cond], sumrank[cond]/Count);
		}
	for (cond = 0; cond < Nconds; cond++)
		R += sumrank[cond] * sumrank[cond];
	chi_r = 12.0 / (Count*Nconds*(Nconds+1)) * R - (3 * Count * (Nconds+1));
	statheader ("Friedman Chi-Square Test for Ranks");
	tprint ("Chi-square of ranks", chi_r);
	chiprint (chi_r, Nconds-1, pochisq (chi_r, Nconds - 1));
	if (Nconds == 2) /* really, this is a two-tailed wilcoxon */
		needtable = TRUE;
	if (Nconds == 3 && Count <= 9)
		needtable = TRUE;
	if (Nconds == 4 && Count <= 4)
		needtable = TRUE;
	if (needtable)
		checktable ("Friedman", "N", Count);
	}

/*FUNCTION dospearman: compute & print rank-order correlation matrix */
/*
	WARNING: this procedure clobbers the original global data
*/	
dospearman ()
	{
	/* the following two vectors save allocation time for ranksort */
	float	*sorted;    /* sorted condition values */
	int 	*order;     /* sorted order index numbers */
	double	r;
	int 	cond;
	int 	cond2;
	
	order = myalloc (int, Count);
	sorted = myalloc (float, Count);
	if (order == NULL || sorted == NULL)
		ERRSPACE(Spearman Rho)
	for (cond = 0; cond < Nconds; cond++)
		if (ranksort (Condat[cond], sorted, order, Count))
			ERRMSG0 (could not rank data for Spearman Rho)
	
	statheader ("Spearman Rank Correlation (rho) [corrected for ties]");
	pcritrho (Count);
	if (Count < NSPEARMAN)
		checktable ("Spearman rho", "N", Count);
	if (Nconds == 2)
		{
		r = cor (Condat[0], Condat[1], Count);
		tprint ("rho", r);
		return;
		}
	
#define	LFORMAT    "%-8.8s "
#define	SFORMAT    "%8.8s "
#define	PFORMAT    "%8.3f "
	printf (LFORMAT, "");
	for (cond = 0; cond < Nconds; cond++)
		printf (SFORMAT, Condname[cond]);
	putchar ('\n');
	for (cond = 0; cond < Nconds; cond++)
		{
		printf (SFORMAT, Condname[cond]);
		for (cond2 = 0; cond2 < Nconds; cond2++)
			{
			if (cond == cond2)
				printf (SFORMAT, "");
			else
				{
				r = cor (Condat[cond], Condat[cond2], Count);
				printf (PFORMAT, r);
				}
			}
		putchar ('\n');
		}
	printf (LFORMAT, "");
	for (cond = 0; cond < Nconds; cond++)
		printf (SFORMAT, Condname[cond]);
	putchar ('\n');
	
	free ((char *) sorted);
	free ((char *) order);
	}

/*FUNCTION critcor: compute critical value for correlation */
/*
	t = r * sqrt (df / (1-r*r));
	F = r*r * df / (1-r*r)
	r = sqrt (F / (df + F))
*/
double
critcor (n, p)
int 	n;   /* number of points */
double	p;   /* significance level */
	{
	int 	df = n - 2;
	double	critf ();         /* critical F ratio */
	double	sqrt ();
	double	r;
	double	f;
	
	if (p <= 0.0 || p >= 1.0 || df < 1)
		return (0.0);
	f = critf (p, 1, df);
	r = sqrt (f / (df + f));
	return (r);
	}

pcritrho (n)
int 	n;
	{
	double	p05 = critcor (n, .05);
	double	p01 = critcor (n, .01);
	tprint ("Critical r (.05) t approximation", p05);
	tprint ("Critical r (.01) t approximation", p01);
	}
